Scaffolding transcoding and docker env

opadron opened the pull request:
Adds more of the initial scaffolding, but more importantly are two other major contributions in this PR:

1) The beginnings of a docker-based girder-worker job for handling the transcoding of source videos to webm+ogg vorbis.

2) A comprehensive turn-key development environment based on docker.

To try this for yourself, see `DEVELOPERS.md` and let me know if you run into any problems.

Commit:
wip

Commit:
add processing docker image/implement POST item/{id}/video

Commit:
update processing route to return original job if ran multiple times

Commit:
implement video processing task

Commit:
improve docker-based dev env

Commit:
nearly complete docker setup

Commit:
add instructions for devs

opadron commented on docker/docker-compose.yml:
```diff
@@ -0,0 +1,91 @@
+  mq:
+    image: rabbitmq
+    ports:
+      - '5672:5672'
```
Note to self: I think we can remove this `ports` section.

opadron commented on docker/ffmpeg_local/convert.py:
```diff
@@ -0,0 +1,117 @@
+
+
+subprocess.call(['ls', GIRDER_WORKER_DIR])
+sys.stdout.flush()
```
Forgot to remove this.

danlamanna commented on docker/build/girder-post-install.py:
```diff
@@ -0,0 +1,143 @@
+               dict(key='worker.api_url', value=apiUrl)])))
+
+client.put('system/restart')
+
```
I'm unsure if you want to add ansible into the mix, but this file largely duplicates what the ansible-client for Girder aims to achieve. Documentation for it exists [here](https://github.com/girder/girder/blob/v2.2.0/devops/ansible/roles/girder/library/README.md).

It could also help replace the sleep with a more appropriate [wait_for](http://docs.ansible.com/ansible/latest/wait_for_module.html) call.

opadron replied:
I had considered using the ansible client, but I thought there are some operations I needed for which there is not an idempotent equivalent, so I decided to keep everything in a single script.  I made a note to investigate adding these operations to the ansible-client.

opadron replied:
After reviewing the ansible client in more depth, I've come to realize that I was mistaken, and that it did, in fact, cover my needed use cases!  I ran into a minor problem, but I think I've got it covered in girder/girder#2310.

@danlamanna, @kotfic et al: well done putting together a pretty sweet Ansible module. 👏

danlamanna commented on docker/docker-compose.yml:
```diff
@@ -0,0 +1,91 @@
+  worker: {}
+services:
+  docker:
+    image: 'docker:dind'
```
@manthey is the resident expert on docker in/alongside docker, but [dind is considered obsolete](https://github.com/jpetazzo/dind#another-word-of-warning) as far as I know.

opadron replied:
So, my interpretation of the same sources leads me to believe that it is not docker-in-docker, itself, that is obsolete; but the old instructions on how to achieve such a setup. ("This work" referring to the repo, not the official "docker" docker image).

"If you want to run Docker-in-Docker today, all you need to do is:"

implies to me that docker-in-docker is still very much supported, but "this is how you do it, now..."

I admit the language and context is vague.

danlamanna commented on docker/ffmpeg_local/convert.py:
```diff
@@ -0,0 +1,117 @@
+
+input_file = next(glob.iglob(os.path.join(GIRDER_WORKER_DIR, 'input.*')))
+
+cmd = [FFMPEG, '-i', input_file]
```
There are also a [couple](https://github.com/kkroening/ffmpeg-python) [of](https://github.com/Ch00k/ffmpy) ffmpeg python bindings out there that could be useful.

opadron replied:
I had also considered bindings for ffmpeg, but forewent them mainly because I don't think they buy us anything technically.  It's probably also worth mentioning that according to the legal understanding of several folks (admittedly, none who are lawyers), using some bindings might expose us to additional claims stemming from the precise definition of "linking" and/or "derivative works".  I, for one, disagree that there is any such distinction, but at least going command-line only gives these folks one less thing to worry about.

danlamanna commented on server/constants.py:
```diff
@@ -28,3 +28,22 @@ class PluginSettings:
+
+
+class JobStatus:
+    """Deferred loading of Girder's JobStatus constants"""
```
Why defer these?

opadron replied:
The objects in this scope are evaluated/initialized on plugin load, where I can't yet take for granted that the code and objects for dependent plugins are initialized and ready for use.

Indeed, I actually tested this and found that my plugin fails to load because
```Python
girder.plugins.jobs.constants
```
fails to import.

So, I have to defer the import until after the plugin is loaded, which other plugins, like large_image, have done.   I just tried to do it in a way I thought was more pythonic; wrapping the details in a proxy object that lazy-imported only when necessary, and well after plugin load.

jcfr commented:
Nice :+1: 

I also like the detailed instructions explaining how to setup the dev env.

May be you could add a `License` section to the `README.rst` referencing the Apache license file already added and briefly describe the caveats related to `ffmpeg` ?

Commit:
switch to girder ansible client

Commit:
minor cleanup

opadron commented:
@jcfr I went ahead and added a copy of the Apache license to the `LICENSE` file.  There is already a section in the `README.rst` that should reference the license terms.

As far as the patents related to ffmpeg go, I'd like to take the approach of restricting the plugin only to features that we know (or are at least very confident) are safe from patent encumbrance issues.  We can revisit this approach at a later time should we choose to expose such a restricted code path.

Edit: I just noticed I already had the `LICENSE` file, so I removed the superfluous commit.

opadron commented:
@manthey ready for review!

manthey commented:
In running this on my local development system, I get the following error during the `docker-compose up -d --build` step:
```
ERROR: for docker_pre-build_1  Cannot start service pre-build: oci runtime error: container_linux.go:262: starting container process caused "process_linux.go:339: container init caused \"rootfs_linux.go:57: mounting \\\"/home/ubuntu/girder/plugins/girder/devops/ansible/roles/girder/library/girder.py\\\" to rootfs \\\"/var/lib/docker/aufs/mnt/04bd689097f4a0433b6385be6f44ac74b8c70499366361008f52257532dbdb97\\\" at \\\"/var/lib/docker/aufs/mnt/04bd689097f4a0433b6385be6f44ac74b8c70499366361008f52257532dbdb97/etc/ansible/roles/girder.girder/library/girder.py\\\" caused \\\"not a directory\\\"\""
: Are you trying to mount a directory onto a file (or vice-versa)? Check if the specified host path exists and is the expected type
```

opadron commented:
@manthey thanks for taking a look at this.  I will investigate your issue, as well as a number of new problems I had also ran into.  I will let you know when this will be ready, again.

Commit:
fix several issues

 - added documentation about rebuilding Docker
   images with "--no-cache"

 - added ".env" file correctly setting the
   COMPOSE_PROJECT_NAME setting

 - fixed an issue preventing the ansible
   "post-install" script from accessing the
   "girder" ansible library module

opadron commented:
@manthey give this another shot.  The error you saw was a mistake on my part that should be fixed, now.

manthey commented:
I spun this up in a Vagrant VM where I had previous run a copy of Girder.  If port 8080 is already in use, the `docker-compose up` step fails without a clear error message, so we probably need to add that you mustn't have that port already in use.  I stopped my old version of girder, did a docker-compose rm, and then docker-compose up.

I uploaded an mp4 and tried to kick off the conversion process; the DIND docker had failed to start, so I removed all the containers, images, and volumes, then rebuilt.

I reuploaded my video and again tried to start the conversion process.  This failed to create the job, throwing an error on line 1251 of model_base.py because the `'access'` key didn't exist in whatever document it was working on.  Have you seen this error?  I uploaded the file to admin public folder.

I added a guard in model_base to only try to move existing keys and let girder restart.  Now the job started and ran successfully.

opadron commented:
@manthey Yeah, I've seen that and several similar errors that have left me puzzled.  What I ended up doing was pulling the latest version of the `girder/girder` Docker image, rebuilding the containers with `--no-cache`, and just for good measure, removing the docker volumes with `docker-compose down --volumes`.  That seems to fix all the strange errors like the one you are seeing.

opadron commented:
Regarding the port 8080 error message, I'll add something to the documentation for the next PR.  I'll have it ready for you to look at, soon!

Commit:
improve convert script/fix minor issue in video endpoint

Commit:
add note about port 8080 to DEVELOPERS

opadron commented:
@manthey PTAL

Commit:
added several minor improvements

  - fixed a typo in DEVLOPERS.md

  - updated docker-compose.yml

    - renamed girder volume and container to "web"

    - renamed mongo container to "db"

    - renamed pre-build container to "provision"
      to better convey its purpose

    - renamed core-build container to "corewatch"

    - renamed video-build container to
      "videowatch" (still commented out, for now)

  - updated video transcoding command line in
    convert.py

    - changed video bitrate from 100K to 1M

    - added an option to enable the use of
      multiple threads

opadron commented:
@manthey this should have the finishing touches.
